pkg/auth/requests/impersonate.go

```diff
@@ -53,23 +58,41 @@ func (h *impersonatingAuth) Authenticate(req *http.Request) (k8sUser.Info, bool,
 			if err != nil {
 				return nil, false, err
 			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate")
+				return nil, false, errors.New("not allowed to impersonate user")
 			}
 			impersonateUser = true
 		}
 
-		if len(reqGroup) > 0 && !groupsEqual(reqGroup, groups) {
-			canDo, err := h.sar.UserCanImpersonateGroups(req, user, reqGroup)
+		if len(reqGroup) > 0 {
+			for _, g := range reqGroup {
+				if slices.Contains(groups, g) {
+					//user belongs to the group they are trying to impersonate
+					continue
+				}
+				canDo, err := h.sar.UserCanImpersonateGroup(req, user, g)
+				if err != nil {
+					return nil, false, err
+				} else if !canDo {
+					return nil, false, errors.New("not allowed to impersonate group")
+				}
+				impersonateGroup = true
+			}
+		}
+
+		if reqExtras != nil && len(reqExtras) > 0 {
```

Comment:
This can be simplified:

```suggestion
		if len(reqExtras) > 0 {
```